archive/imo/imo2008_q4.lean

```diff
@@ -0,0 +1,135 @@
+/-
+Copyright (c) 2021 Manuel Candales. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Manuel Candales
+-/
+import data.real.basic
+import data.real.sqrt
+
+/-!
+# IMO 2008 Q4
+Find all functions `f : (0,∞) → (0,∞)` (so, `f` is a function from the positive real
+numbers to the positive real numbers) such that
+      ```
+      (f(w)^2 + f(x)^2)/(f(y^2) + f(z^2)) = (w^2 + x^2)/(y^2 + z^2)
+      ```
+for all positive real numbers `w`, `x`, `y`, `z`, satisfying `wx = yz`.
+
+# Solution
+The desired theorem is that either `f = λ x, x` or `f = λ x, 1/x`
+-/
+
+open real
+
+theorem imo2008_q4
+  (f : ℝ → ℝ)
+  (H₁ : ∀ x > 0, f(x) > 0) :
+  (∀ w x y z : ℝ, 0 < w → 0 < x → 0 < y → 0 < z → w*x = y*z →
+    (f(w)^2 + f(x)^2)/(f(y^2) + f(z^2)) = (w^2 + x^2)/(y^2 + z^2)) ↔
+  ((∀ x > 0, f(x) = x) ∨ (∀ x > 0, f(x) = 1/x)) :=
+begin
+  split, swap,
+  -- proof that f(x) = x and f(x) = 1/x satisfy the condition
+  { rintros (h | h),
+    { intros w x y z hw hx hy hz hprod,
+      rw [(h w hw), (h x hx), (h (y^2) (pow_pos hy 2)), (h (z^2) (pow_pos hz 2))] },
+    { intros w x y z hw hx hy hz hprod,
+      rw [(h w hw), (h x hx), (h (y^2) (pow_pos hy 2)), (h (z^2) (pow_pos hz 2))],
+      have hy2z2 : y^2 + z^2 ≠ 0 := ne_of_gt (add_pos (pow_pos hy 2) (pow_pos hz 2)),
+      have hz2y2 : z^2 + y^2 ≠ 0 := ne_of_gt (add_pos (pow_pos hz 2) (pow_pos hy 2)),
+      have hp2 : w^2 * x^2 = y^2 * z^2, { rw [(mul_pow w x 2).symm, (mul_pow y z 2).symm, hprod] },
+      field_simp [(ne_of_gt hw), (ne_of_gt hx), (ne_of_gt hy), (ne_of_gt hz), hy2z2, hz2y2, hp2],
```

Comment:
```suggestion
      field_simp [ne_of_gt hw, ne_of_gt hx, ne_of_gt hy, ne_of_gt hz, hy2z2, hz2y2, hp2],
```
If you want you can remove all the parentheses around elements of lists (also in many other places of the proof). I personally find that more readable, but if you think it's less readable, then you can keep it as-is.